File: PostalCode/ViewController/FavoriteViewController.m
//
//  FavoriteViewController.m
//  PostalCode
//
//  Created by OCHIISHI Koichiro on 10/10/13.
//  Copyright (c) 2013 OCHIISHI Koichiro. All rights reserved.
//

#import "FavoriteViewController.h"

@interface FavoriteViewController ()

@property (nonatomic, strong) NSMutableArray *objects;
@property (nonatomic, strong) UIBarButtonItem *deleteButtonItem;

@end

@implementation FavoriteViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.tableView.estimatedRowHeight = 44.f;
    self.tableView.rowHeight = UITableViewAutomaticDimension;
    
    self.deleteButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"消去"
                                                             style:UIBarButtonItemStylePlain
                                                            target:self
                                                            action:@selector(deleteAllData)];
    self.navigationItem.leftBarButtonItem = self.deleteButtonItem;

    [self reloadAllData];
}

- (void)viewWillAppear:(BOOL)animated
{
    // UITableViewController は clearsSelectionOnViewWillAppear により、
    // super 時に、選択中の Row を非選択状態に戻すため、super 前に選択中の Row があるかを取得しておく
    bool hasSelectedRow = self.tableView.indexPathForSelectedRow != nil;
    [super viewWillAppear:animated];

    // self.tableView deselectRowAtIndexPath: animated: を有効にする為に、
    // 他 UITabBarController.viewControllers から選択された場合のみ、reload 処理を行う
    if (!hasSelectedRow) {
        [self reloadAllData];
    }
    
    self.deleteButtonItem.enabled = (self.objects.count) ? YES : NO;
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark -

- (void)reloadAllData
{
    self.objects = [FavoriteRepository getFavorites];
    [self.tableView reloadData];
}

- (void)deleteAllData
{
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:nil message:nil preferredStyle:UIAlertControllerStyleActionSheet];

    [alertController addAction:[UIAlertAction actionWithTitle:@"すべての項目を削除" style:UIAlertActionStyleDestructive handler:^(UIAlertAction *action) {
        [FavoriteRepository deleteAllFavorite];
        [self reloadAllData];
        self.deleteButtonItem.enabled = NO;
    }]];
    
    [alertController addAction:[UIAlertAction actionWithTitle:@"キャンセル" style:UIAlertActionStyleCancel handler:nil]];
    
    [self presentViewController:alertController animated:YES completion:nil];
}

- (NSString *)textLabelText:(NSIndexPath *)indexPath
{
    NSData *data = self.objects[indexPath.row];
    PostalCodeModel *model = (PostalCodeModel *)[NSKeyedUnarchiver unarchiveObjectWithData:data];
    return [NSString stringWithFormat:@"%@ %@ %@", model.stateK, model.cityTownK, model.streetK];
}

- (NSString *)detailTextLabelText:(NSIndexPath *)indexPath
{
    NSData *data = self.objects[indexPath.row];
    PostalCodeModel *model = (PostalCodeModel *)[NSKeyedUnarchiver unarchiveObjectWithData:data];
    
    NSMutableString *postalCode = [NSMutableString stringWithString:model.postalCode];
    [postalCode insertString:@"-" atIndex:3];
    
    return  [postalCode copy];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.objects.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *CellIdentifier = @"Cell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier forIndexPath:indexPath];
    
    // Configure the cell...
    cell.textLabel.text = [self textLabelText:indexPath];
    cell.detailTextLabel.text = [self detailTextLabelText:indexPath];
    
    return cell;
}

// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath
{
    // Return NO if you do not want the specified item to be editable.
    return YES;
}

// Override to support editing the table view.
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        // Delete the row from the data source
        [tableView beginUpdates];

        [FavoriteRepository deleteFavoritePostalCodeModel:indexPath.row];
        
        [self.objects removeObjectAtIndex:indexPath.row];
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationLeft];
        
        [tableView endUpdates];
    }
    else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
    }   
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSData *data = self.objects[indexPath.row];
    PostalCodeModel *model = (PostalCodeModel *)[NSKeyedUnarchiver unarchiveObjectWithData:data];
    
    DetailViewController *viewController = [self.storyboard instantiateViewControllerWithIdentifier:@"DetailViewController"];
    viewController.postalCodeModel = model;
    [self.navigationController pushViewController:viewController animated:YES];
}

@end

File: PostalCode/ViewController/FavoriteViewController.swift
//
//  FavoriteViewController.swift
//  PostalCode
//
//  Created by Koichiro OCHIISHI on 2025/05/05.
//  Copyright © 2025 OCHIISHI Koichiro. All rights reserved.
//

import UIKit

class FavoriteViewController: BaseTableViewController {

    private var postalCodes: [PostalCode] = []
    private lazy var deleteButtonItem: UIBarButtonItem = {
        UIBarButtonItem(
            title: "消去",
            style: .plain,
            target: self,
            action: #selector(deleteAllData)
        )
    }()

    override func viewDidLoad() {
        super.viewDidLoad()

        tableView.estimatedRowHeight = 44.0
        tableView.rowHeight = UITableView.automaticDimension

        navigationItem.leftBarButtonItem = deleteButtonItem

        reloadAllData()
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)

        let hasSelectedRow = tableView.indexPathForSelectedRow != nil

        if !hasSelectedRow {
            reloadAllData()
        }

        deleteButtonItem.isEnabled = !postalCodes.isEmpty
    }

    // MARK: - Data Handling

    private func reloadAllData() {
        postalCodes = FavoriteRepository.getFavorites()
        tableView.reloadData()
    }

    @objc private func deleteAllData() {
        let alertController = UIAlertController(
            title: nil, message: nil, preferredStyle: .actionSheet)

        alertController.addAction(
            UIAlertAction(title: "すべての項目を削除", style: .destructive) { _ in
                FavoriteRepository.deleteAllFavorite()
                self.reloadAllData()
                self.deleteButtonItem.isEnabled = false
            })

        alertController.addAction(UIAlertAction(title: "キャンセル", style: .cancel, handler: nil))

        present(alertController, animated: true, completion: nil)
    }

    private func textLabelText(for postalCode: PostalCode) -> String {
        return "\(postalCode.stateK) \(postalCode.cityTownK) \(postalCode.streetK)"
    }

    private func detailTextLabelText(for indexPath: IndexPath) -> String {
        let postalCode = postalCodes[indexPath.row]
        return postalCode.formattedCode
    }

    // MARK: - Table View Data Source

    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return postalCodes.count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath)
        -> UITableViewCell
    {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath)
        let postalCode = self.postalCodes[indexPath.row]

        cell.textLabel?.text = textLabelText(for: postalCode)
        cell.detailTextLabel?.text = detailTextLabelText(for: indexPath)

        return cell
    }

    // MARK: - Table View Delegate

    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        return true
    }

    override func tableView(
        _ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle,
        forRowAt indexPath: IndexPath
    ) {
        guard editingStyle == .delete else { return }

        tableView.beginUpdates()

        FavoriteRepository.deleteFavorite(at: indexPath.row)
        postalCodes.remove(at: indexPath.row)
        tableView.deleteRows(at: [indexPath], with: .left)

        tableView.endUpdates()
    }

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let postalCode = self.postalCodes[indexPath.row]

        let viewController =
            storyboard?.instantiateViewController(withIdentifier: "DetailViewController")
            as! DetailViewController
        viewController.postalCode = postalCode
        navigationController?.pushViewController(viewController, animated: true)
    }
}
